### PR TITLE
strokeWidth as a channel

### LINKED ISSUE #311: strokeWidth channel on links
would be useful to draw a graph (I thought we already had an issue for this?)

### PR DESCRIPTION
**Pull request by Fil**
closes #311

build : https://observablehq.com/@fil/plot-strokewidth-322

### CODE REVIEW

**Comment by Fil** [comment]
Using this for a Sankey diagram, I'll need to set the width with a scale: identity. (Doesn't mean a scale isn't going to be useful in general.)

**Comment by ryoqun** [comment]
@Fil @mbostock hi, is this going to be merged soon? :)

Also, I want tick's strokeWidth to be a channel as well.

**Comment by Fil** [comment]
See #175 and #380

**Review by mbostock — CHANGES_REQUESTED** [review]
I support not having a scale for strokeWidth (at least for now), but it does feel arbitrary to only expose this property on the link mark. I would prefer to make this a standard channel on all marks, like we do with strokeOpacity. Even if that channel often goes unused, it’s still meaningful, and should be supported for consistency/symmetry. I might take a pass at this now.

**Comment by ryoqun** [comment]
> I support not having a scale for strokeWidth (at least for now), but it does feel arbitrary to only expose this property on the link mark. I would prefer to make this a standard channel on all marks, like we do with strokeOpacity. Even if that channel often goes unused, it’s still meaningful, and should be supported for consistency/symmetry. I might take a pass at this now.

hey, I've always been fan of d3.js. thanks for moving this forward. :)

File: README.md [hunk 1]
```diff
@@ -459,7 +459,6 @@ All marks support the following style options:
 * **fill** - fill color
 * **fillOpacity** - fill opacity (a number between 0 and 1)
 * **stroke** - stroke color
-* **strokeWidth** - stroke width (in pixels)
```
Comment (mbostock) [inline README.md:462]: This should be kept, right? The options that are supported as channels are also listed as constant options.

**Comment by ryoqun** [comment]
hooray! thanks for working hard on this. so, strokeWidth for all marks? cool. :)

fyi, I'm creating a nice chart like this with plot (simply put, we're debugging any performance issues for distributed computing node's consensus formulation):

image

And I want to use `strokeWidth` to embed extra scalar values into marks, instead of radius for visual conciseness and proper visual comparison (radius/area isn't cognitively easy than just width for visual measuring).

**Comment by Fil** [comment]
wow! great example, I'd be curious to see it in action if you can share it! If you want to test this branch I've updated the build at https://observablehq.com/@fil/plot-strokewidth-322 (see also https://observablehq.com/@fil/plot-early-bird)

**Comment by ryoqun** [comment]
>  I'd be curious to see it in action if you can share it!

yeah, I'd like to share but it's still in bad shape to share currently... needs some preparation to be published to the public..

> If you want to test this branch I've updated the build at https://observablehq.com/@fil/plot-strokewidth-322 (see also https://observablehq.com/@fil/plot-early-bird)

seems to work nicely! (I applied strokeWidth to Plot.dot and Plot.tickX. I can't wait official new version release with it. :)

image

File: README.md [hunk 2]
```diff
@@ -476,7 +476,7 @@ All marks support the following optional channels:
 * **strokeOpacity** - a stroke opacity; bound to the *opacity* scale
```
Comment (mbostock) [inline README.md:476]: Add strokeWidth here?

**Review by mbostock — APPROVED** [review]

**Comment by Fil** [comment]
example for CHANGELOG:
strokeWidth-channel

Plot.dotX(d3.range(41), { strokeWidth: (d) => (1 + d) / 15 }).plot()

**Comment by enjalot** [comment]
> Can do, but I'm not sure it's worth it. Links tend to come with value=1…10 (co-occurences as in the miserables.json example), or 0…1, or sometimes from distance matrices 0…infty, and I feel it's always more "graphic tweaking" than really using it as an encoding to be shared with another mark.

I didn't think about the "shared with another mark" aspect, but from the perspective of "Plot figures out how to map data space to graphic space for me and I just want to tweak the pixels" having a scale here is what I expect. It's similar to how I would think about the `r` channel in dots, where I usually want to impose some maximum size. That could be considered a graphic tweak but it's a pretty meaningful one.
I suppose for the same reason I also want a scale for fontSize, because without a scale I have to go outside of Plot to calculate the extent of the data and do the math that Plot would do for me in the other channels, like opacity

**Pull request merged**